Course/Course/Presentation/Dates/CourseDatesViewModel.swift

```diff
@@ -149,14 +200,221 @@ extension CourseDatesViewModel {
             Task {
                 await getCourseDates(courseID: courseID)
                 await MainActor.run { [weak self] in
-                    self?.dueDatesShifted = true
+                    guard let self else { return }
+                    eventState = .shiftedDueDates
                 }
             }
         }
     }
     
     func resetDueDatesShiftedFlag() {
-        dueDatesShifted = false
+        eventState = CourseDatesViewModel.EventState.none
+    }
+}
+
+extension CourseDatesViewModel {
+    private func handleCalendar() {
+        calendar.requestAccess { [weak self] _, previousStatus, status in
+            guard let self else { return }
+            switch status {
+            case .authorized:
+                showAddCalendarAlert()
+            default:
+                isOn = false
+                if previousStatus == status {
+                    self.showCalendarSettingsAlert()
+                }
+            }
+        }
+    }
+    
+    @MainActor
+    func addCourseEvents(trackAnalytics: Bool = true, completion: ((Bool) -> Void)? = nil) {
+        guard let dateBlocks = courseDates?.dateBlocks else { return }
+        showCalendarSyncProgressView { [weak self] in
+            guard let self else { return }
+            calendar.addEventsToCalendar(for: dateBlocks) { [weak self] calendarEventsAdded in
+                guard let self else { return }
+                self.isOn = calendarEventsAdded
+                if calendarEventsAdded {
+                    self.calendar.syncOn = calendarEventsAdded
+                    self.router.dismiss(animated: false)
+                    self.showEventsAddedSuccessAlert()
+                }
+                completion?(calendarEventsAdded)
+            }
+        }
+    }
+    
+    func removeCourseCalendar(trackAnalytics: Bool = true, completion: ((Bool) -> Void)? = nil) {
+        calendar.removeCalendar { [weak self] success in
+            guard let self else { return }
+            self.isOn = !success
+            completion?(success)
+        }
+    }
+    
+    private func showAddCalendarAlert() {
+        router.presentAlert(
+            alertTitle: CourseLocalization.CourseDates.addCalendarTitle,
+            alertMessage: CourseLocalization.CourseDates.addCalendarPrompt(
+                config.platformName,
+                courseName
+            ),
+            positiveAction: CoreLocalization.Alert.accept,
+            onCloseTapped: { [weak self] in
+                guard let self else { return }
```

Comment:
It seems strong reference if `self` isn't required in this type of blocks because if the self will be nill, by using the optional self will not crash the app and following code will automatically be ignored.